### DISCUSSION TITLE
I have some confusion regarding the transactions used inside msg_insert function in PostgreSQL.

### DISCUSSION THREAD

**Discussion by shein-haha** [FerretDB/FerretDB#2659] [General]
Why is the msg_insert function in PostgreSQL using a single transaction to perform single-row inserts for all MongoDB insert operations? When we conducted some basic performance tests, we found that putting each insert operation into a transaction had a negative impact on performance. FerretDB converts multi-row MongoDB inserts into single-row inserts in PostgreSQL, so I believe it is unnecessary to include PostgreSQL's single-row inserts in a transaction.

**Comment by AlekSi** [comment]
Your analysis is mostly correct. However, it is a bit more complicated when some documents in the middle of the batch can't be inserted: both the `ordered: true` and `ordered: false` behaviors of the `insert` command do not match PostgreSQL transaction semantics, so we have to handle them differently. That was the original reason for the current implementation. The good news is that it is already optimized in the coming release, and more optimizations will be made in the coming months.

**Reply by shein-haha** [reply]
Thank you very much for your answer!
